src/client/serialize/hdf_serialize.c

```diff
@@ -0,0 +1,1034 @@
+		rc = 1;
+		goto out;
+	}
+	status = H5Tset_size(attr_dtype, strlen(entry->dpe_str) + 1);
+	if (status < 0) {
+		fprintf(stderr, "failed to set dtype size");
+		rc = 1;
+		goto out;
+	}
+	status = H5Tset_strpad(attr_dtype, H5T_STR_NULLTERM);
+	if (status < 0) {
+		fprintf(stderr, "failed to set null terminator");
+		rc = 1;
+		goto out;
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		fprintf(stderr, "failed to create version attribute dataspace");
+		rc = 1;
+		goto out;
+	}
+
+	usr_attr = H5Acreate2(*file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		fprintf(stderr, "failed to create attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Awrite(usr_attr, attr_dtype, entry->dpe_str);
+	if (status < 0) {
+		fprintf(stderr, "failed to write attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Aclose(usr_attr);
+	if (status < 0) {
+		fprintf(stderr, "failed to close attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Tclose(attr_dtype);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dtype");
+		rc = 1;
+		goto out;
+	}
+	status = H5Sclose(attr_dspace);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dspace");
+		rc = 1;
+		goto out;
+	}
+out:
+	return rc;
+}
+
+static int
+cont_serialize_prop_uint(hid_t *file_id,
+			 struct daos_prop_entry *entry,
+			 const char *prop_str)
+{
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype;
+	hid_t	attr_dspace;
+	hid_t	usr_attr;
+
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_NATIVE_UINT64);
+	status = H5Tset_size(attr_dtype, 8);
+	if (status < 0) {
+		fprintf(stderr, "failed to create version dtype");
+		rc = 1;
+		goto out;
+	}
+	if (attr_dtype < 0) {
+		fprintf(stderr, "failed to create usr attr type");
+		rc = 1;
+		goto out;
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		fprintf(stderr, "failed to create version attr dspace");
+		rc = 1;
+		goto out;
+	}
+	usr_attr = H5Acreate2(*file_id, prop_str, attr_dtype,
+			      attr_dspace, H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		fprintf(stderr, "failed to create attr");
+		rc = 1;
+		goto out;
+	}
+	status = H5Awrite(usr_attr, attr_dtype, &entry->dpe_val);
+	if (status < 0) {
+		fprintf(stderr, "failed to write attr");
+		rc = 1;
+		goto out;
+	}
+	status = H5Aclose(usr_attr);
+	if (status < 0) {
+		fprintf(stderr, "failed to close attr");
+		rc = 1;
+		goto out;
+	}
+	status = H5Tclose(attr_dtype);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dtype");
+		rc = 1;
+		goto out;
+	}
+	status = H5Sclose(attr_dspace);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dspace");
+		rc = 1;
+		goto out;
+	}
+out:
+	return rc;
+}
+
+static int
+cont_serialize_props(hid_t *file_id, daos_prop_t *prop_query,
+		     daos_handle_t cont)
+{
+	int			rc = 0;
+	struct daos_prop_entry	*entry;
+
+	entry = &prop_query->dpp_entries[0];
```

Comment:
how are you going to guarantee that whenever one adds a new property, they are going to remember that they add the serializing here?
can we do something like a loop to get all container properties and serialize them vs hardcoding each one?

Reply:
@mchaarawi we have a list of the container properties serialized documented in the milestone documentation for serialization. I suppose I didn't realize that the container properties would change very often. I also had the initial thought of serializing each container property as its own attribute so that the name of the attribute was the container property name, and the attribute value was the container property value. I suppose we could serialize the container properties similarly to the way that the user attributes are serialized, which is done using an hdf5 compound datatype for usr attribute name and the usr attribute value, which allows varying lengths and datatypes. I could create a container property compound datatype that contains a container property name and a container property value. Is this the better option?

Reply:
@mchaarawi also, there is this issue now: https://jira.hpdd.intel.com/browse/DAOS-7765 where every POSIX container should have a unique label. So is it best to serialize/deserialize the container label property? Because if someone serializes then deserializes a container and an old one still exists then the container label will not be unique?

Reply:
as for the label, yes we should serialize it. if a container with the same label exists in the pool where we are deserializing, we should fail, or provider the option to change the label.